core/src/main/scala/org/allenai/common/cache/QueryCache.scala

```diff
@@ -0,0 +1,93 @@
+package org.allenai.common.cache
+
+import java.io._
+import java.nio.ByteBuffer
+import java.nio.charset.Charset
+
+import com.lambdaworks.redis.RedisClient
+import com.lambdaworks.redis.codec.RedisCodec
+import org.allenai.common.Logging
+
+/** trait holding a Redis cache of query results. This is meant to store any value T where
+  * T <: Object, keyed on string query.
+  * Only one syncQueryCache per application, unlimited AsyncQueryCaches.
+  * @param redisHostname name of host where redis server is running
+  * @param redisHostPort port of redis server on host
+  * @param clientPrefix an identifier for the client using this caching mechanism, which will become
+  * part of the cache key (prepended to the actual query)
+  */
+abstract class QueryCache[V <: Any](redisHostname: String, redisHostPort: Int, clientPrefix: String)
+    extends Logging {
+
+  val redisClient = new RedisClient(redisHostname, redisHostPort)
+  /** Trivial Helper to construct cache key with client prefix.
+    */
+  protected def keyForQuery(query: String): String = s"${clientPrefix}_$query"
+
+  /** Retrieves the value for a passed key
+    * @param query key for stored value (not including client prefix)
+    * @return Option containing value, None if not found or timed out (async)
+    */
+  def get(query: String): Option[V]
+
+  /** Puts a key->value to the cache
+    * @param query key for value (not including client prefix)
+    * @param response Value you want stored in cache
+    */
+  def put(query: String, response: V): Unit
+
+  /** deltes a key->value from the cache
+    * @param query key for value you want to delete (not including client prefix)
+    */
+  def del(query: String): Unit
+}
+
+class Boxing[T]
```

Comment:
You should have each class in its own file here, and add a bit of documentation.